File: apps/mobile/ios/Runner/TdlibBridge.swift
import Flutter
import UIKit
import Foundation

@_silgen_name("td_json_client_create")
private func td_json_client_create() -> OpaquePointer?

@_silgen_name("td_json_client_send")
private func td_json_client_send(_ client: OpaquePointer?, _ request: UnsafePointer<CChar>?)

@_silgen_name("td_json_client_receive")
private func td_json_client_receive(_ client: OpaquePointer?, _ timeout: Double) -> UnsafePointer<CChar>?

@_silgen_name("td_json_client_destroy")
private func td_json_client_destroy(_ client: OpaquePointer?)

class TdlibBridge: NSObject {
    private let channel: FlutterMethodChannel
    private let queue = DispatchQueue(label: "spnetgram.tdlib")
    private var client: OpaquePointer?
    private var apiId: Int = 0
    private var apiHash: String = ""
    private var receiving = false
    private var pendingChats: Set<Int64> = []
    private var chatCache: [Int64: [String: Any]] = [:]
    private let databasePath: String

    init(controller: FlutterViewController) {
        self.channel = FlutterMethodChannel(name: "spnetgram/tdlib", binaryMessenger: controller.binaryMessenger)
        let documents = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first
        self.databasePath = documents?.appendingPathComponent("tdlib").path ?? "tdlib"
        super.init()
        self.channel.setMethodCallHandler(handle)
    }

    private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            if let args = call.arguments as? [String: Any] {
                apiId = args["apiId"] as? Int ?? 0
                apiHash = args["apiHash"] as? String ?? ""
            }
            ensureClient()
            result(true)
        case "sendPhoneNumber":
            if let args = call.arguments as? [String: Any] {
                let phone = args["phone"] as? String ?? ""
                send(["@type": "setAuthenticationPhoneNumber", "phone_number": phone])
            }
            result(true)
        case "sendOtp":
            if let args = call.arguments as? [String: Any] {
                let code = args["code"] as? String ?? ""
                send(["@type": "checkAuthenticationCode", "code": code])
            }
            result(true)
        case "sendPassword":
            if let args = call.arguments as? [String: Any] {
                let password = args["password"] as? String ?? ""
                send(["@type": "checkAuthenticationPassword", "password": password])
            }
            result(true)
        case "registerUser":
            if let args = call.arguments as? [String: Any] {
                let firstName = args["firstName"] as? String ?? ""
                let lastName = args["lastName"] as? String ?? ""
                send(["@type": "registerUser", "first_name": firstName, "last_name": lastName])
            }
            result(true)
        case "fetchChats":
            send(["@type": "getChats", "chat_list": ["@type": "chatListMain"], "limit": 50])
            result(true)
        case "fetchMessages":
            if let args = call.arguments as? [String: Any] {
                let chatId = args["chatId"] as? Int64 ?? 0
                send([
                    "@type": "getChatHistory",
                    "chat_id": chatId,
                    "from_message_id": 0,
                    "offset": 0,
                    "limit": 30,
                    "only_local": false
                ])
            }
            result(true)
        case "sendMessage":
            if let args = call.arguments as? [String: Any] {
                let chatId = args["chatId"] as? Int64 ?? 0
                let text = args["text"] as? String ?? ""
                send([
                    "@type": "sendMessage",
                    "chat_id": chatId,
                    "input_message_content": [
                        "@type": "inputMessageText",
                        "text": ["@type": "formattedText", "text": text],
                        "disable_web_page_preview": true,
                        "clear_draft": true
                    ]
                ])
            }
            result(true)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func ensureClient() {
        if client != nil { return }
        client = td_json_client_create()
        startReceiveLoop()
    }

    private func startReceiveLoop() {
        if receiving { return }
        receiving = true
        queue.async { [weak self] in
            guard let self = self else { return }
            while self.receiving {
                if let response = td_json_client_receive(self.client, 1.0) {
                    let json = String(cString: response)
                    self.handleTdlibResponse(json)
                }
            }
        }
    }

    private func send(_ payload: [String: Any]) {
        guard let client = client else { return }
        if payload["@type"] as? String == "setTdlibParameters" {
            // no-op
        }
        let data = try? JSONSerialization.data(withJSONObject: payload, options: [])
        guard let jsonData = data, let jsonString = String(data: jsonData, encoding: .utf8) else { return }
        jsonString.withCString { ptr in
            td_json_client_send(client, ptr)
        }
    }

    private func handleTdlibResponse(_ json: String) {
        guard let data = json.data(using: .utf8) else { return }
        guard let obj = try? JSONSerialization.jsonObject(with: data, options: []),
              let dict = obj as? [String: Any],
              let type = dict["@type"] as? String else { return }

        switch type {
        case "updateAuthorizationState":
            if let state = dict["authorization_state"] as? [String: Any],
               let stateType = state["@type"] as? String {
                if stateType == "authorizationStateWaitTdlibParameters" {
                    send([
                        "@type": "setTdlibParameters",
                        "parameters": [
                            "@type": "tdlibParameters",
                            "database_directory": databasePath,
                            "use_message_database": true,
                            "use_file_database": true,
                            "use_secret_chats": false,
                            "api_id": apiId,
                            "api_hash": apiHash,
                            "system_language_code": "en",
                            "device_model": UIDevice.current.model,
                            "system_version": UIDevice.current.systemVersion,
                            "application_version": "1.0",
                            "enable_storage_optimizer": true
                        ]
                    ])
                } else if stateType == "authorizationStateWaitPhoneNumber" {
                    emit("onAuthState", ["status": "WAIT_PHONE"])
                } else if stateType == "authorizationStateWaitCode" {
                    emit("onAuthState", ["status": "WAIT_CODE"])
                } else if stateType == "authorizationStateWaitPassword" {
                    emit("onAuthState", ["status": "WAIT_PASSWORD"])
                } else if stateType == "authorizationStateWaitEncryptionKey" {
                    send(["@type": "checkDatabaseEncryptionKey", "encryption_key": ""])
                } else if stateType == "authorizationStateWaitOtherDeviceConfirmation" {
                    let link = state["link"] as? String ?? ""
                    emit("onAuthState", ["status": "WAIT_DEVICE_CONFIRMATION", "message": link])
                } else if stateType == "authorizationStateWaitRegistration" {
                    emit("onAuthState", ["status": "WAIT_REGISTRATION"])
                } else if stateType == "authorizationStateReady" {
                    emit("onAuthState", ["status": "READY"])
                    send(["@type": "getChats", "chat_list": ["@type": "chatListMain"], "limit": 50])
                } else {
                    emit("onAuthState", ["status": "UNKNOWN"])
                }
            }
        case "chats":
            if let ids = dict["chat_ids"] as? [Int64] {
                pendingChats = Set(ids)
                chatCache.removeAll()
                ids.forEach { id in
                    send(["@type": "getChat", "chat_id": id])
                }
                if ids.isEmpty {
                    emit("onChats", [])
                }
            }
        case "chat":
            if let chatId = dict["id"] as? Int64 {
                let title = dict["title"] as? String ?? "Chat"
                let lastMessage = extractLastMessage(dict["last_message"] as? [String: Any])
                let unread = dict["unread_count"] as? Int ?? 0
                chatCache[chatId] = [
                    "id": chatId,
                    "title": title,
                    "lastMessage": lastMessage,
                    "unreadCount": unread
                ]
                pendingChats.remove(chatId)
                if pendingChats.isEmpty {
                    emit("onChats", Array(chatCache.values))
                }
            }
        case "messages":
            if let msgs = dict["messages"] as? [[String: Any]] {
                for msg in msgs.reversed() {
                    emitMessage(msg)
                }
            }
        case "updateNewMessage":
            if let msg = dict["message"] as? [String: Any] {
                emitMessage(msg)
            }
        default:
            break
        }
    }

    private func emitMessage(_ msg: [String: Any]) {
        let chatId = msg["chat_id"] as? Int64 ?? 0
        let id = msg["id"] as? Int64 ?? 0
        let date = msg["date"] as? Int ?? 0
        let isOutgoing = msg["is_outgoing"] as? Bool ?? false
        let sender = msg["sender_id"] as? [String: Any]
        let senderLabel: String
        if let senderType = sender?["@type"] as? String {
            if senderType == "messageSenderUser" {
                senderLabel = "user:\(sender?["user_id"] as? Int ?? 0)"
            } else if senderType == "messageSenderChat" {
                senderLabel = "chat:\(sender?["chat_id"] as? Int ?? 0)"
            } else {
                senderLabel = "unknown"
            }
        } else {
            senderLabel = "unknown"
        }

        let text = extractMessageText(msg["content"] as? [String: Any])
        emit("onMessage", [
            "id": id,
            "chatId": chatId,
            "sender": senderLabel,
            "text": text,
            "timestamp": date,
            "isOutgoing": isOutgoing
        ])
    }

    private func extractLastMessage(_ msg: [String: Any]?) -> String {
        guard let msg = msg else { return "" }
        return extractMessageText(msg["content"] as? [String: Any])
    }

    private func extractMessageText(_ content: [String: Any]?) -> String {
        guard let content = content,
              let type = content["@type"] as? String else { return "" }
        if type == "messageText", let text = content["text"] as? [String: Any] {
            return text["text"] as? String ?? ""
        }
        return ""
    }

    private func emit(_ method: String, _ args: Any?) {
        DispatchQueue.main.async {
            self.channel.invokeMethod(method, arguments: args)
        }
    }
}

File: apps/legacy-mobile/android/app/src/main/kotlin/com/spnetgram/app/TdlibBridge.kt
package com.spnetgram.app

import android.os.Handler
import android.os.Looper
import android.util.Log
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.drinkless.tdlib.Client
import org.drinkless.tdlib.TdApi
import java.util.concurrent.atomic.AtomicInteger

class TdlibBridge(private val engine: FlutterEngine) : MethodChannel.MethodCallHandler {
    private val channel = MethodChannel(engine.dartExecutor.binaryMessenger, "spnetgram/tdlib")
    private val mainHandler = Handler(Looper.getMainLooper())
    private var client: Client? = null
    private var apiId: Int = 0
    private var apiHash: String = ""
    private var authorizationState: TdApi.AuthorizationState? = null

    fun register() {
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> {
                apiId = call.argument<Int>("apiId") ?: 0
                apiHash = call.argument<String>("apiHash") ?: ""
                ensureClient()
                result.success(true)
            }
            "sendPhoneNumber" -> {
                val phone = call.argument<String>("phone") ?: ""
                send(TdApi.SetAuthenticationPhoneNumber(phone, null))
                result.success(true)
            }
            "sendOtp" -> {
                val code = call.argument<String>("code") ?: ""
                send(TdApi.CheckAuthenticationCode(code))
                result.success(true)
            }
            "sendPassword" -> {
                val password = call.argument<String>("password") ?: ""
                send(TdApi.CheckAuthenticationPassword(password))
                result.success(true)
            }
            "registerUser" -> {
                val firstName = call.argument<String>("firstName") ?: ""
                val lastName = call.argument<String>("lastName") ?: ""
                send(TdApi.RegisterUser(firstName, lastName))
                result.success(true)
            }
            "fetchChats" -> {
                fetchChats()
                result.success(true)
            }
            "fetchMessages" -> {
                val chatId = call.argument<Long>("chatId") ?: 0L
                fetchMessages(chatId)
                result.success(true)
            }
            "sendMessage" -> {
                val chatId = call.argument<Long>("chatId") ?: 0L
                val text = call.argument<String>("text") ?: ""
                sendMessage(chatId, text)
                result.success(true)
            }
            else -> result.notImplemented()
        }
    }

    private fun ensureClient() {
        if (client != null) return
        client = Client.create(updateHandler, null, null)
    }

    private fun send(function: TdApi.Function) {
        client?.send(function, defaultHandler)
    }

    private val defaultHandler = Client.ResultHandler { obj ->
        if (obj is TdApi.Error) {
            Log.e("TDLIB", "TDLib error: ${obj.message}")
        }
    }

    private val updateHandler = Client.ResultHandler { obj ->
        when (obj) {
            is TdApi.UpdateAuthorizationState -> {
                authorizationState = obj.authorizationState
                handleAuthState(obj.authorizationState)
            }
            is TdApi.UpdateNewMessage -> {
                emitMessage(obj.message)
            }
            else -> {
            }
        }
    }

    private fun handleAuthState(state: TdApi.AuthorizationState) {
        when (state) {
            is TdApi.AuthorizationStateWaitTdlibParameters -> {
                val params = TdApi.TdlibParameters().apply {
                    databaseDirectory = engine.context.filesDir.absolutePath + "/tdlib"
                    filesDirectory = engine.context.filesDir.absolutePath + "/tdlib_files"
                    useMessageDatabase = true
                    useFileDatabase = true
                    useSecretChats = false
                    apiId = this@TdlibBridge.apiId
                    apiHash = this@TdlibBridge.apiHash
                    systemLanguageCode = "en"
                    deviceModel = android.os.Build.MODEL
                    systemVersion = android.os.Build.VERSION.RELEASE
                    applicationVersion = "1.0"
                    enableStorageOptimizer = true
                }
                send(TdApi.SetTdlibParameters(params))
            }
            is TdApi.AuthorizationStateWaitEncryptionKey -> {
                send(TdApi.CheckDatabaseEncryptionKey(ByteArray(0)))
            }
            is TdApi.AuthorizationStateWaitPhoneNumber -> emitAuth("WAIT_PHONE")
            is TdApi.AuthorizationStateWaitCode -> emitAuth("WAIT_CODE")
            is TdApi.AuthorizationStateWaitPassword -> emitAuth("WAIT_PASSWORD")
            is TdApi.AuthorizationStateWaitOtherDeviceConfirmation -> emitAuth(
                "WAIT_DEVICE_CONFIRMATION",
                state.link
            )
            is TdApi.AuthorizationStateWaitRegistration -> emitAuth("WAIT_REGISTRATION")
            is TdApi.AuthorizationStateReady -> {
                emitAuth("READY")
                fetchChats()
            }
            is TdApi.AuthorizationStateLoggingOut -> emitAuth("LOGGING_OUT")
            is TdApi.AuthorizationStateClosed -> emitAuth("CLOSED")
            else -> emitAuth("UNKNOWN")
        }
    }

    private fun fetchChats() {
        val request = TdApi.GetChats(TdApi.ChatListMain(), 50)
        client?.send(request) { obj ->
            if (obj is TdApi.Chats) {
                val ids = obj.chatIds
                if (ids.isEmpty()) {
                    emitChats(emptyList())
                    return@send
                }
                val remaining = AtomicInteger(ids.size)
                val chats = mutableListOf<TdApi.Chat>()
                ids.forEach { chatId ->
                    client?.send(TdApi.GetChat(chatId)) { chatObj ->
                        if (chatObj is TdApi.Chat) {
                            synchronized(chats) { chats.add(chatObj) }
                        }
                        if (remaining.decrementAndGet() == 0) {
                            emitChats(chats)
                        }
                    }
                }
            }
        }
    }

    private fun fetchMessages(chatId: Long) {
        val request = TdApi.GetChatHistory(chatId, 0, 0, 30, false)
        client?.send(request) { obj ->
            if (obj is TdApi.Messages) {
                obj.messages?.reversed()?.forEach { msg ->
                    emitMessage(msg)
                }
            }
        }
    }

    private fun sendMessage(chatId: Long, text: String) {
        val content = TdApi.InputMessageText(TdApi.FormattedText(text, null), false, true)
        val request = TdApi.SendMessage(chatId, 0, 0, null, null, content)
        send(request)
    }

    private fun emitAuth(status: String, message: String? = null) {
        val payload = mutableMapOf<String, Any>("status" to status)
        if (message != null) {
            payload["message"] = message
        }
        emit("onAuthState", payload)
    }

    private fun emitChats(list: List<TdApi.Chat>) {
        val payload = list.map { chat ->
            mapOf(
                "id" to chat.id,
                "title" to chat.title,
                "lastMessage" to lastMessageText(chat.lastMessage),
                "unreadCount" to chat.unreadCount
            )
        }
        emit("onChats", payload)
    }

    private fun emitMessage(message: TdApi.Message) {
        val payload = mapOf(
            "id" to message.id,
            "chatId" to message.chatId,
            "sender" to senderLabel(message.senderId),
            "text" to messageText(message.content),
            "timestamp" to message.date,
            "isOutgoing" to message.isOutgoing
        )
        emit("onMessage", payload)
    }

    private fun emit(method: String, args: Any?) {
        mainHandler.post { channel.invokeMethod(method, args) }
    }

    private fun messageText(content: TdApi.MessageContent?): String {
        return when (content) {
            is TdApi.MessageText -> content.text.text
            else -> ""
        }
    }

    private fun senderLabel(sender: TdApi.MessageSender?): String {
        return when (sender) {
            is TdApi.MessageSenderUser -> "user:${sender.userId}"
            is TdApi.MessageSenderChat -> "chat:${sender.chatId}"
            else -> "unknown"
        }
    }

    private fun lastMessageText(message: TdApi.Message?): String {
        if (message == null) return ""
        return messageText(message.content)
    }
}
